code/cfile/cfilesystem.cpp

```diff
@@ -72,62 +72,65 @@ typedef struct cf_root_sort {
 	int				cf_type;
 } cf_root_sort;
 
-#define CF_NUM_ROOTS_PER_BLOCK   32
-#define CF_MAX_ROOT_BLOCKS			256				// Can store 32*256 = 8192 Roots
-#define CF_MAX_ROOTS					(CF_NUM_ROOTS_PER_BLOCK * CF_MAX_ROOT_BLOCKS)
+#define CF_NUM_ROOTS_PER_BLOCK		32
+#define CF_MIN_ROOT_BLOCKS			2				// minimum number of root blocks to create on resize
 
 typedef struct cf_root_block {
 	cf_root			roots[CF_NUM_ROOTS_PER_BLOCK];
 } cf_root_block;
 
 static int Num_roots = 0;
-static cf_root_block  *Root_blocks[CF_MAX_ROOT_BLOCKS];
+static SCP_vector<cf_root_block *> Root_blocks;
 
 static int Num_path_roots = 0;
 
 // Created by searching all roots in order.   This means Files is then sorted by precedence.
 typedef struct cf_file {
-	char		name_ext[CF_MAX_FILENAME_LENGTH];	// Filename and extension
+	SCP_string	name_ext;	// Filename and extension
 	int			root_index;							// Where in Roots this is located
 	int			pathtype_index;						// Where in Paths this is located
 	time_t		write_time;							// When it was last written
 	int			size;								// How big it is in bytes
 	int			pack_offset;						// For pack files, where it is at.   0 if not in a pack file.  This can be used to tell if in a pack file.
-	char*		real_name;							// For real files, the full path
+	SCP_string	real_name;							// For real files, the full path
 	const void*	data;								// For in-memory files, the data pointer
+
+	cf_file() : root_index(0), pathtype_index(0), write_time(0), size(0), pack_offset(0), data(nullptr) {}
 } cf_file;
 
-#define CF_NUM_FILES_PER_BLOCK   512
-#define CF_MAX_FILE_BLOCKS			128						// Can store 512*128 = 65536 files
+#define CF_NUM_FILES_PER_BLOCK		512
+#define CF_MIN_FILE_BLOCKS			16				// minimum number of file blocks to create on resize
 
 typedef struct cf_file_block {
 	cf_file						files[CF_NUM_FILES_PER_BLOCK];
 } cf_file_block;
 
 static uint Num_files = 0;
-static cf_file_block  *File_blocks[CF_MAX_FILE_BLOCKS];
+static SCP_vector<cf_file_block *> File_blocks;
```

Comment:
You could use `unique_ptr`s to avoid the manual memory management here.

Reply:
Good point. I don't really use them enough for that to be my goto for stuff like this.